File: backend/schemas/player.py
from __future__ import annotations

from datetime import datetime
from pydantic import BaseModel


# ---- Team ----
class TeamOut(BaseModel):
    id: int
    name: str
    name_en: str | None = None
    abbreviation: str | None = None
    conference: str | None = None
    division: str | None = None
    logo_url: str | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


# ---- Player Stats ----
class PlayerStatsOut(BaseModel):
    id: int
    player_id: int
    season: str
    team_abbreviation: str | None = None
    gp: int = 0
    gs: int = 0
    mpg: float | None = None
    ppg: float | None = None
    rpg: float | None = None
    apg: float | None = None
    spg: float | None = None
    bpg: float | None = None
    topg: float | None = None
    pfpg: float | None = None
    fg_pct: float | None = None
    three_pct: float | None = None
    ft_pct: float | None = None

    model_config = {"from_attributes": True}


# ---- Player ----
class PlayerSummary(BaseModel):
    id: int
    nba_id: int
    name: str
    name_en: str | None = None
    team: TeamOut | None = None
    jersey_number: str | None = None
    position: str | None = None
    height: str | None = None
    weight: str | None = None
    country: str | None = None
    college: str | None = None
    draft_year: int | None = None
    draft_round: int | None = None
    draft_pick: int | None = None
    years_exp: int | None = None
    image_url: str | None = None
    latest_stats: PlayerStatsOut | None = None
    is_active: bool = True

    model_config = {"from_attributes": True}


class PlayerDetail(PlayerSummary):
    career_stats: list[PlayerStatsOut] = []


class PlayerComparisonItem(BaseModel):
    id: int
    name: str
    name_en: str | None = None
    team: TeamOut | None = None
    image_url: str | None = None
    stats: PlayerStatsOut | None = None

    model_config = {"from_attributes": True}


class PlayerComparisonOut(BaseModel):
    players: list[PlayerComparisonItem]


# ---- Pagination ----
class PaginationMeta(BaseModel):
    page: int
    page_size: int
    total: int
    total_pages: int


class PaginatedResponse(BaseModel):
    data: list
    meta: PaginationMeta


# ---- API Envelope ----
class ApiResponse(BaseModel):
    data: list | dict | None = None
    meta: dict | None = None


# ---- Scrape ----
class ScrapeTriggerRequest(BaseModel):
    source: str  # "players" or "stats"


class ScrapeLogOut(BaseModel):
    id: int
    source: str
    status: str
    records_added: int
    records_updated: int
    error_message: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}

File: frontend/src/types/player.ts
export interface Team {
  id: number
  name: string
  name_en: string | null
  abbreviation: string | null
  conference: string | null
  division: string | null
  logo_url: string | null
}

export interface PlayerStats {
  id: number
  player_id: number
  season: string
  team_abbreviation: string | null
  gp: number
  gs: number
  mpg: number | null
  ppg: number | null
  rpg: number | null
  apg: number | null
  spg: number | null
  bpg: number | null
  topg: number | null
  pfpg: number | null
  fg_pct: number | null
  three_pct: number | null
  ft_pct: number | null
}

export interface PlayerSummary {
  id: number
  nba_id: number
  name: string
  name_en: string | null
  team: Team | null
  jersey_number: string | null
  position: string | null
  height: string | null
  weight: string | null
  country: string | null
  college: string | null
  draft_year: number | null
  draft_round: number | null
  draft_pick: number | null
  years_exp: number | null
  image_url: string | null
  latest_stats: PlayerStats | null
  is_active: boolean
}

export interface PlayerDetail extends PlayerSummary {
  career_stats: PlayerStats[]
}

export interface PlayerComparisonItem {
  id: number
  name: string
  name_en: string | null
  team: Team | null
  image_url: string | null
  stats: PlayerStats | null
}

export interface PlayerComparisonOut {
  players: PlayerComparisonItem[]
}

export interface PlayerSearchResult {
  id: number
  name: string
  name_en: string | null
  team: Team | null
  position: string | null
}

export interface PaginationMeta {
  page: number
  page_size: number
  total: number
  total_pages: number
}

export interface ApiResponse<T> {
  data: T
  meta?: PaginationMeta
}
